File: solutions/javascript/clock/1/clock.js
export class Clock {
  constructor(hours = 0, minutes = 0) {
    this.totalMinutes = this.normalize(hours * 60 + minutes);
  }

  normalize(minutes) {
    const minutesInDay = 24 * 60;
    return ((minutes % minutesInDay) + minutesInDay) % minutesInDay;
  }

  toString() {
    const hours = Math.floor(this.totalMinutes / 60);
    const minutes = this.totalMinutes % 60;
    return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}`;
  }

  plus(minutesToAdd) {
    return new Clock(0, this.totalMinutes + minutesToAdd);
  }

  minus(minutesToSubtract) {
    return new Clock(0, this.totalMinutes - minutesToSubtract);
  }

  equals(otherClock) {
    return this.totalMinutes === otherClock.totalMinutes;
  }
}

File: solutions/swift/clock/1/Sources/Clock/Clock.swift
import Foundation

struct Clock: Equatable {
    private let totalMinutes: Int
    
 
    private init(totalMinutes: Int) {
        let minutesInDay = 24 * 60

        let normalizedMinutes = (totalMinutes % minutesInDay + minutesInDay) % minutesInDay
        self.totalMinutes = normalizedMinutes
    }
    

    init(hours: Int, minutes: Int = 0) {
        self.init(totalMinutes: hours * 60 + minutes)
    }
    

    var description: String {
        let hours = totalMinutes / 60
        let minutes = totalMinutes % 60
        return String(format: "%02d:%02d", hours, minutes)
    }
    

    func add(minutes: Int) -> Clock {
        return Clock(totalMinutes: self.totalMinutes + minutes)
    }
    

    func subtract(minutes: Int) -> Clock {
        return Clock(totalMinutes: self.totalMinutes - minutes)
    }
    

}
